File: src/generic/octree.cpp
#include <pcl/point_cloud.h>
#include <pcl/octree/octree.h>

#define Octree_ptr pcl::octree::OctreePointCloudSearch<_PointT>::Ptr
#define PointCloud_ptr pcl::PointCloud<_PointT>::Ptr

PCLIMP(void*, OctreePointCloudSearch, new)(double resolution)
{
  return new Octree_ptr(new pcl::octree::OctreePointCloudSearch<_PointT>(resolution));
}

PCLIMP(void, OctreePointCloudSearch, delete)(Octree_ptr *self)
{
  delete self;
}

PCLIMP(double, OctreePointCloudSearch, getResolution)(Octree_ptr *self)
{
  return (*self)->getResolution();
}

PCLIMP(double, OctreePointCloudSearch, getEpsilon)(Octree_ptr *self)
{
  return (*self)->getEpsilon();
}

PCLIMP(void, OctreePointCloudSearch, setEpsilon)(Octree_ptr *self, double value)
{
  (*self)->setEpsilon(value);
}

PCLIMP(void, OctreePointCloudSearch, setInputCloud)(Octree_ptr *self, PointCloud_ptr *cloud, Indices_ptr *indices)
{
  (*self)->setInputCloud(*cloud, indices ? *indices : pcl::octree::OctreePointCloudSearch<_PointT>::IndicesConstPtr());
}

PCLIMP(void, OctreePointCloudSearch, addPointsFromInputCloud)(Octree_ptr *self)
{
  (*self)->addPointsFromInputCloud();
}

PCLIMP(void, OctreePointCloudSearch, addPointToCloud)(Octree_ptr *self, const _PointT &point, PointCloud_ptr *cloud)
{
  (*self)->addPointToCloud(point, *cloud);
}

PCLIMP(bool, OctreePointCloudSearch, isVoxelOccupiedAtPoint)(Octree_ptr *self, const _PointT &point)
{
  return (*self)->isVoxelOccupiedAtPoint(point);
}

PCLIMP(void, OctreePointCloudSearch, deleteTree)(Octree_ptr *self)
{
  (*self)->deleteTree();
}

PCLIMP(void, OctreePointCloudSearch, setMaxVoxelIndex)(Octree_ptr *self, unsigned int value)
{
  (*self)->setMaxVoxelIndex(value);
}

PCLIMP(void, OctreePointCloudSearch, setTreeDepth)(Octree_ptr *self, unsigned int value)
{
  (*self)->setTreeDepth(value);
}

PCLIMP(unsigned int, OctreePointCloudSearch, getTreeDepth)(Octree_ptr *self)
{
  return (*self)->getTreeDepth();
}

PCLIMP(unsigned int, OctreePointCloudSearch, getLeafCount)(Octree_ptr *self)
{
  return static_cast<unsigned int>((*self)->getLeafCount());
}

PCLIMP(unsigned int, OctreePointCloudSearch, getBranchCount)(Octree_ptr *self)
{
  return static_cast<unsigned int>((*self)->getBranchCount());
}

PCLIMP(int, OctreePointCloudSearch, nearestKSearch)(Octree_ptr *self, const _PointT &point, int k, Indices_ptr *indices, THFloatTensor *squaredDistances)
{
  std::vector<int> dummy;
  std::vector<int>& indices_ = (indices && *indices) ? **indices : dummy;
  std::vector<float> squaredDistances_;
  
  int found = (*self)->nearestKSearch(point, k, indices_, squaredDistances_);
    
  if (squaredDistances)
    vector2Tensor(squaredDistances_, squaredDistances);

  return found;
}

PCLIMP(int, OctreePointCloudSearch, radiusSearch)(Octree_ptr *self, const _PointT &point, double radius, Indices_ptr *indices, THFloatTensor *squaredDistances, unsigned int max_nn)
{
  std::vector<int> dummy;
  std::vector<int>& indices_ = (indices && *indices) ? **indices : dummy;
  std::vector<float> squaredDistances_;
  
  int found = (*self)->radiusSearch(point, radius, indices_, squaredDistances_, max_nn);

  if (squaredDistances)
    vector2Tensor(squaredDistances_, squaredDistances);  
  
  return found;
}

#undef Octree_ptr
#undef PointCloud_ptr 

File: Octree.lua
local ffi = require 'ffi'
local torch = require 'torch'
local utils = require 'pcl.utils'
local pcl = require 'pcl.PointTypes'

local Octree = torch.class('pcl.Octree', pcl)

local func_by_type = {}

local function init()

  local OctreePointCloudSearch_method_names = {
    'new',
    'delete',
    'getResolution',
    'getEpsilon',
    'setEpsilon',
    'setInputCloud',
    'addPointsFromInputCloud',
    'addPointToCloud',
    'isVoxelOccupiedAtPoint',
    'deleteTree', 
    'setMaxVoxelIndex',
    'setTreeDepth',
    'getTreeDepth',
    'getLeafCount',
    'getBranchCount',
    'nearestKSearch',
    'radiusSearch'
  }

  for k,v in pairs(utils.type_key_map) do
    func_by_type[k] = utils.create_typed_methods("pcl_OctreePointCloudSearch_TYPE_KEY_", OctreePointCloudSearch_method_names, v)
  end    
end

init()

function Octree:__init(pointType, resolution)
  if type(pointType) == 'number' then
    resolution = pointType
    pointType = pcl.PointXYZ
  end
  
  utils.check_arg('resolution', type(resolution) == 'number', 'resolution not specified')
  
  pointType = pcl.pointType(pointType)
  rawset(self, 'f', func_by_type[pointType])
  self.o = self.f.new(resolution)
end

function Octree:cdata()
  return self.o
end

function Octree:getResolution()
  return self.f.getResolution(self.o)
end

function Octree:getEpsilon()
  return self.f.getEpsilon(self.o)
end

function Octree:setEpsilon(eps)
  self.f.setEpsilon(self.o, eps)
end

function Octree:setInputCloud(cloud)
  self.f.setInputCloud(self.o, cloud:cdata())
end

function Octree:addPointsFromInputCloud()
  self.f.addPointsFromInputCloud(self.o)
end

function Octree:addPointToCloud(point, cloud)
  self.f.addPointToCloud(self.o, point, cloud:cdata())
end

function Octree:isVoxelOccupiedAtPoint(p)
  self.f.isVoxelOccupiedAtPoint(self.o, p)
end

function Octree:deleteTree()
  self.f.deleteTree(self.o)
end

function Octree:setMaxVoxelIndex(max_index)
  self.f.setMaxVoxelIndex(self.o, max_index)
end

function Octree:setTreeDepth(depth)
  return self.f.setTreeDepth(self.o, depth)
end

function Octree:getTreeDepth()
  return self.f.getTreeDepth(self.o)
end

function Octree:getLeafCount()
  return self.f.getLeafCount(self.o)
end

function Octree:getBranchCount()
  return self.f.getBranchCount(self.o)
end

function Octree:nearestKSearch(point, k, indices, squaredDistances)
  return self.f.nearestKSearch(self.o, point, k, utils.opt(indices), utils.opt(squaredDistances))
end

function Octree:radiusSearch(point, radius, indices, squaredDistances, max_nn)
  return self.f.radiusSearch(self.o, point, radius, utils.opt(indices), utils.opt(squaredDistances), max_nn or 0)
end
